tests/pos/i7807.scala

```diff
@@ -0,0 +1,3 @@
+object Test with
+
+  def flip: (x: 0 | 1) => x.type match { case 0 => 1 case 1 => 0 } = ???
```

Comment:
I would add some more tests:
```scala
   flip(0): 1
   flip(1): 0

   flip(if ??? then 0 else 1)
   val n: 0 | 1 = if ??? then 0 else 1
   flip(n)

   val m: n.type match { case 0 => 1 case 1 => 0 } = flip(n)
```

But I noticed a strange thing: a match type is apparently not a subtype of the union of it branches. So adding `flip(m)` or `flip(flip(n))` for instance does not work:

```scala
-- [E007] Type Mismatch Error: tests/pos/i7807.scala:15:8 --------------------------------------------------------------
15 |   flip(m)
   |        ^
   |        Found:    (Test.m :
   |          (Test.n : (0 : Int) | (1 : Int)) match {
   |            case (0 : Int) => (1 : Int)
   |            case (1 : Int) => (0 : Int)
   |          }
   |        )
   |        Required: (0 : Int) | (1 : Int)
```

Should an issue be opened about this? Though it seems the current spec does not say this should work, I find it surprising that it does not.

Reply:
I am not sure it should work. Generally, match types are on a thin line between expressiveness and soundness. So generally demanding they do more than they do would have to come with an algorithm that shows how to do it and an argument why this would not break soundness.